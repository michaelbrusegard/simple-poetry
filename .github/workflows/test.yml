name: Test

on:
  push:
    branches:
      - main

jobs:
  test:
    name: Run Tests
    runs-on: self-hosted
    steps:
      - name: Checkout code
        uses: actions/checkout@v4

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: '22'

      - name: Install dependencies
        run: npm install

      - name: Run tests
        run: npx vitest run
