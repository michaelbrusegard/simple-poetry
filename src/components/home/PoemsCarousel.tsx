import { useState, useEffect } from 'react'
import styles from './Poems.module.css'
import { useQuery } from '@tanstack/react-query'
import { PoemCard } from './PoemCard'
import { fetchPoems } from './poemsquery'
import { Button } from '../ui/Button'
import { Spinner } from '../ui/Spinner'

function PoemsCarousel({
  searchTerm,
  selectedAuthor,
}: {
  searchTerm: string
  selectedAuthor: string
}) {
  const poems = useQuery({
    queryKey: ['poems', searchTerm, selectedAuthor],
    queryFn: () => fetchPoems(searchTerm, selectedAuthor),
  })

  const [currentIndex, setCurrentIndex] = useState(0)

  useEffect(() => {
    setCurrentIndex(0)
  }, [searchTerm, selectedAuthor])

  const prevPoem = () => {
    setCurrentIndex((oldIndex) => Math.max(0, oldIndex - 1))
  }

  const nextPoem = () => {
    setCurrentIndex((oldIndex) => Math.min(poems.data.length - 1, oldIndex + 1))
  }

  if (poems.error) {
    return <p className={styles.errorMessage}>Error: {poems.error.message}</p>
  }

  if (
    !poems.isLoading &&
    (!Array.isArray(poems.data) || poems.data.length === 0)
  ) {
    return <p className={styles.errorMessage}>No poems found...</p>
  }

  return (
    <div className={styles.carousel}>
      <Button
        className={styles.carouselButton}
        variant='outline'
        size='icon'
        onClick={prevPoem}
        disabled={currentIndex === 0}
      >
        Prev
      </Button>
      {poems.isLoading ? (
        <Spinner />
      ) : (
        <PoemCard
          key={poems.data[currentIndex].title}
          className={styles.carouselPoem}
          poem={poems.data[currentIndex]}
        />
      )}
      <Button
        className={styles.carouselButton}
        variant='outline'
        size='icon'
        onClick={nextPoem}
        disabled={poems.isLoading || currentIndex === poems.data.length - 1}
      >
        Next
      </Button>
    </div>
  )
}

export { PoemsCarousel }
