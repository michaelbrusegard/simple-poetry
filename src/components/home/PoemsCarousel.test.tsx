import { screen, fireEvent, act, waitFor } from '@testing-library/react'
import { PoemsCarousel } from './PoemsCarousel'
import { render } from '../../test/utils'

describe('PoemsCarousel', () => {
  it('should render the poems and navigate between them', async () => {
    render(
      <PoemsCarousel searchTerm='Love' selectedAuthor='William Shakespeare' />,
    )

    await waitFor(() => {
      expect(screen.getByText("A Lover's Complaint")).toBeInTheDocument()
    })

    fireEvent.click(screen.getByText('Next'))

    await waitFor(() => {
      expect(
        screen.getByText(
          "Sonnet 10: For shame! deny that thou bear'st love to any",
        ),
      ).toBeInTheDocument()
    })

    fireEvent.click(screen.getByText('Prev'))

    await waitFor(() => {
      expect(screen.getByText("A Lover's Complaint")).toBeInTheDocument()
    })
  })

  it('should match the snapshot', async () => {
    let fragment
    await act(async () => {
      const { asFragment } = render(
        <PoemsCarousel
          searchTerm='Love'
          selectedAuthor='William Shakespeare'
        />,
      )
      fragment = asFragment()
    })
    expect(fragment).toMatchSnapshot()
  })
})
