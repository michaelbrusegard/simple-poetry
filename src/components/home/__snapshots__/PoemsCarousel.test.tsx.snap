// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PoemsCarousel > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="_carousel_276d41"
  >
    <button
      class="_button_8ea813 _variantOutline_8ea813 _sizeIcon_8ea813 _carouselButton_276d41"
      disabled=""
    >
      Prev
    </button>
    <a
      class="_carouselPoem_276d41"
      href="/poem/William Shakespeare;A Lover's Complaint"
    >
      <div
        class="_card_204ad1 _card_871016"
      >
        <div
          class="_cardHeader_204ad1 "
        >
          <h2
            class="_cardTitle_204ad1 "
          >
            A Lover's Complaint
          </h2>
        </div>
        <div
          class="_cardContent_204ad1 "
        >
          William Shakespeare
        </div>
        <div
          class="_cardFooter_204ad1 "
        >
          331 lines
        </div>
      </div>
    </a>
    <button
      class="_button_8ea813 _variantOutline_8ea813 _sizeIcon_8ea813 _carouselButton_276d41"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;
